### ISSUE TITLE
Create a branch for Docker-based Drupal Pi

### ISSUE THREAD

**Issue by geerlingguy** [geerlingguy/drupal-pi#36]
As the title says... Basically I've converted the Raspberry Pi Dramble (http://www.pidramble.com) over to Kubernetes, and now I'd like to do the same for this project, but since it's a single Pi and doesn't have many resources, just use Docker + Docker Compose.

It would simplify the project a bit, and show how to do a single-node Docker-based production-ready Drupal setup.

I would still have Nginx sitting in front, and I would even set it up so I can have my own single-Drupal-Pi route as a proxy to the cluster if needed (so it would basically be a proxy/load balancer/cache in front of the Pi Dramble cluster when it's running as a Kubernetes cluster!).

So some tasks (running list, will be updated as I go):

  - [x] Remove all the web-host-specific bits that will be run as Docker containers
  - [x] Make sure Docker is installed (maybe share the Dramble's [docker-arm](https://github.com/geerlingguy/raspberry-pi-dramble/tree/master/roles/docker-arm) role?)
  - [x] Build a default `example.docker-compose.yml` file to be used with the `geerlingguy/drupal` and `mysql:5.7` images.
  - [x] Add a local Vagrant testing environment.
  - [x] Add 'hooks' which allow task files to be included if they exist.
  - [x] Build a hook that will allow me to install [Drupal for Kubernetes](https://github.com/geerlingguy/drupal-for-kubernetes) on the cluster instead of the default image, by building tasks to clone the project, build the Docker image, and use it instead of `geerlingguy/drupal:latest`.
  - [x] Mount a drupal-files directory from the host into the Drupal container at a configurable path. And make sure to wipe it in the `reset.yml` playbook.
  - [x] Update the `reset.yml` playbook so it destroys the Docker Compose configs and gets Drupal reinstalled from scratch.
  - [x] Add some configuration which can set the Pi to either route traffic from Nginx to the local Docker containers, or to a set of backends in the Pi Dramble cluster for a given URL.

I may or may not make this new branch the default / master, depending on how it performs on a single Pi, and whether it makes the entire user experience _better_ or _worse_.

**Comment by geerlingguy** [comment]
Working on an upstream issue for the docker_arm role... https://github.com/geerlingguy/raspberry-pi-dramble/issues/146

**Comment by geerlingguy** [comment]
Role for Docker on ARM / Raspberry Pi with Ansible is now available: https://galaxy.ansible.com/geerlingguy/docker_arm

**Comment by geerlingguy** [comment]
Almost finished with this. I have a local task hook that is building my Docker for Kubernetes site in the Pi... need to also start testing on a _real Pi_ too, instead of inside Vagrant and Travis CI...

**Comment by geerlingguy** [comment]
Documenting here the `hook-tasks.yml` I'm using to build the 'Drupal for Kubernetes' codebase locally on the Pi for use as the `drupal_image`:

```yaml
---
- name: Create Docker daemon.json config file to fix DNS problems.
  copy:
    content: '{"dns": ["8.8.8.8"]}'
    dest: /etc/docker/daemon.json
  notify: restart docker

- meta: flush_handlers

- name: Install Docker python library.
  pip:
    name: docker
    state: present

- name: Ensure Drupal for Kubernetes project directory exists.
  file:
    path: /opt/drupal-for-kubernetes
    state: directory

- name: Clone Drupal for Kubernetes project.
  git:
    repo: 'https://github.com/geerlingguy/drupal-for-kubernetes.git'
    dest: /opt/drupal-for-kubernetes
    version: master

- name: Build Drupal for Kubernetes Docker image.
  docker_image:
    path: /opt/drupal-for-kubernetes
    name: geerlingguy/drupal-for-kubernetes
    tag: latest
```

**Comment by geerlingguy** [comment]
Almost done with what I need for the next bit of legwork for my Drupal/Kubernetes/Pi presentation!

**Comment by geerlingguy** [comment]
Got the load balancing working in the Vagrant test environment... the next step here is running the entire revamped playbook against my Pi model 2 B+ which is running the live site. Do I 'do it live' and see what breaks and possibly take down www.pidramble.com for a few hours or a day? Or do I do it the boring, safe way, and do it on one of my boring spare Pis to make sure there aren't any more little bugs left to iron out?

Not sure, but too tired to do that tonight. Sounds like a fun thing to do over the weekend!

**Comment by geerlingguy** [comment]
Also bumped into this fun little regex issue: https://github.com/geerlingguy/drupal-container/issues/19

But now that it's resolved, my single pi is all set up and running, woot!

<img width="1368" alt="pidramble" src="https://user-images.githubusercontent.com/481677/53665723-dc623b00-3c31-11e9-992a-953bd953e04e.png">

**Comment by Careless** [comment]
Is there some sort of guide that you have hosted somewhere for linux n00bs like myself to attempt to do this?

I came from the other thread where you mentioned that you will be switching to docker install.

I'm so confused. where do i start, and how do I configure this for just a simple drupal install on a raspberry pi?

There's  dramble-pi, vagrant install, kubernites, docker... not asking you to hold my hand, but the steps to try and get this running should perhaps be in one place. 

Or maybe just a brief list of what to do in what order?

I'm so lost and I'm usually pretty good at figuring my way around things like this but I don't even know where to start or if installing one package a certain way is not going to be compatible with what you're publishing here.

**Comment by Careless** [comment]
geerlingguy, 

I noticed that the guide on the drupal-pi repo was updated to reflect the new docker configuration and gave it a shot, and it seems to be all working now!!!

thanks for publishing that update a few weeks ago. I was confused as I had looked at it earlier and didn't realize that it had been changed to a docker install guide, so now all is well.

I added the VNC ports as suggested in the other thread (5900,5800) and it's all working good.

Now I just need to figure out  a way to migrate my drupal install from one local server (NAS) to another (drupal-pi).

Thanks!
